server/schedulers/balance_region.go

```diff
@@ -179,31 +179,92 @@ func (s *balanceRegionScheduler) transferPeer(cluster schedule.Cluster, region *
 		schedulerCounter.WithLabelValues(s.GetName(), "create_operator_fail").Inc()
 		return nil
 	}
+	s.hitsCounter.miss(source, target)
+	s.hitsCounter.miss(source, nil)
 	return op
 }
 
-// hasPotentialTarget is used to determine whether the specified sourceStore
-// cannot find a matching targetStore in the long term.
-// The main factor for judgment includes StoreState, DistinctScore, and
-// ResourceScore, while excludes factors such as ServerBusy, too many snapshot,
-// which may recover soon.
-func (s *balanceRegionScheduler) hasPotentialTarget(cluster schedule.Cluster, region *core.RegionInfo, source *core.StoreInfo, opInfluence schedule.OpInfluence) bool {
-	filters := []schedule.Filter{
-		schedule.NewExcludedFilter(nil, region.GetStoreIds()),
-		schedule.NewDistinctScoreFilter(cluster.GetLocationLabels(), cluster.GetRegionStores(region), source),
+type record struct {
+	lastTime time.Time
+	count    int
+}
+type hitsStoreBuilder struct {
+	hits      map[string]*record
+	ttl       time.Duration
+	threshold int
+}
+
+func newHitsStoreBuilder(ttl time.Duration, threshold int) *hitsStoreBuilder {
+	return &hitsStoreBuilder{
+		hits:      make(map[string]*record),
+		ttl:       ttl,
+		threshold: threshold,
 	}
+}
 
-	for _, store := range cluster.GetStores() {
-		if schedule.FilterTarget(cluster, store, filters) {
-			continue
-		}
-		if !store.IsUp() || store.DownTime() > cluster.GetMaxStoreDownTime() {
-			continue
+func (h *hitsStoreBuilder) getKey(source, target *core.StoreInfo) string {
+	var key string
+	if source == nil || source == nil && target == nil {
+		return ""
+	} else if target == nil {
+		key = fmt.Sprintf("s%d", source.GetID())
+	} else {
+		key = fmt.Sprintf("s%d->t%d", source.GetID(), target.GetID())
+	}
+	return key
+}
+
+func (h *hitsStoreBuilder) filter(source, target *core.StoreInfo) bool {
+	key := h.getKey(source, target)
+	if item, ok := h.hits[key]; ok && key != "" {
+		if time.Since(item.lastTime) > h.ttl {
+			delete(h.hits, key)
 		}
-		if !shouldBalance(cluster, source, store, region, core.RegionKind, opInfluence) {
-			continue
+		if time.Since(item.lastTime) <= h.ttl && item.count >= h.threshold {
+			return true
 		}
-		return true
 	}
 	return false
 }
+
+func (h *hitsStoreBuilder) miss(source, target *core.StoreInfo) {
+	key := h.getKey(source, target)
+	if _, ok := h.hits[key]; ok && key != "" {
+		delete(h.hits, key)
+	}
+}
+
+func (h *hitsStoreBuilder) hit(source, target *core.StoreInfo) {
```

Comment:
I think the name `hit` and `miss` are really confused here. It is different from the meaning `hit en entry` or `miss an entry` of the cache.
BTW, why not use the `server/cache` directly?

Reply:
Do you have any better suggestion name? `server/cache` not suitable for this situation, we need to build a `Filter` with the `ttl` and hit `count`.

Reply:
Maybe just call them put/remove? @nolouch 

@disksing  How do you think?

Reply:
Seems `put/remove` is better.